packages/build/src/build-and-upload.ts

```diff
@@ -23,26 +23,35 @@ export default async function buildAndUpload(
 
   if (config.dryRun) return;
 
+  for (const key of [
+    'evgAwsKey', 'evgAwsSecret', 'project', 'revision', 'downloadCenterAwsKey', 'downloadCenterAwsSecret'
+  ]) {
+    if (typeof (config as any)[key] !== 'string') {
+      throw new Error(`Missing build config key: ${key}`);
+    }
+  }
+
   // Always release internally to evergreen
   await uploadToEvergreen(
     tarballFile.path,
-    config.evgAwsKey,
-    config.evgAwsSecret,
-    config.project,
-    config.revision
+    config.evgAwsKey as string,
```

Comment:
why is this necessary shouldn't it  be a string in config?

Reply:
It’s defined as an optional property (which is correct, it’s initialized as `process.env.AWS_KEY` which is a string or `undefined`), so we have to explicitly say it’s a string here … we could also make it always a string and default to an empty one in the build config instead, I guess?

Reply:
I mean you only need that if is unchecked. If we "validate" the value for not being undefined should also change the type in subsequent code.  For example:

```
  config: {
    prop?: string;
  } = {};

  const prop1  = config.prop; // here is string or undefined;

  if (!config.prop) {
    throw new Error('error');
  }

  const prop2 = config.prop; // here is only `string`
```

I think in this case it doesn't work cause is done dynamically:

```
 for (const key of [
    'evgAwsKey', 'evgAwsSecret', 'project', 'revision', 'downloadCenterAwsKey', 'downloadCenterAwsSecret'
  ]) {
    if (typeof (config as any)[key] !== 'string') {
      throw new Error(`Missing build config key: ${key}`);
    }
  }
```

Not sure what's best but leaving it as it is is a good option :).

Reply:
So you’d suggest testing all of the keys manually with individual `if`s? I think I would prefer not to do that … :smile:

Reply:
Another solution might be to use the `!` operator, so for example we could use `config.evgAwsKey!` instead of `config.evgAwsKey as string` which makes it easier to read IMHO. What do you think?

Reply:
I avoided that because it leads to a warning by eslint :woman_shrugging: